Clarify error message when non-existing API object update fails with a conflict error because of a precondition

matte21 opened the pull request:
Clarify the error message that is returned when an update precondition on UID or ResourceVersion fails because the updated object does not exist.
More precisely, clearly say that the update failed because the object does not exist rather than mentioning a generic precondition failure.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind documentation

<!--
Add one of the following kinds:
/kind bug
/kind cleanup

/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

**What this PR does / why we need it**:
Clarify an error message that was confusing to users.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #89985

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
The error message that is returned when an API object update fails because of a UID/ResourceVersion precondition failure caused by the fact that the API object does not exist now explicitly mentions that the object does not exist rather than talking about a generic precondition failure.

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

matte21 committed:
Clarify error message

Clarify the error message that is returned when an update precondition
on UID or ResourceVersion fails because the updated object does not
exist.
More precisely, clearly say that the update failed because the object
does not exist rather than mentioning a generic precondition failure.

matte21 commented:
@MikeSpreitzer
@liggitt
@lavalamp

matte21 commented:
/assign @liggitt

liggitt commented:
can you add unit tests that exercise the uid and resourceVersion preconditions in the "not found" and "not matching" cases that demonstrate the error message we expect

fedebongio commented:
/assign @jingyih

matte21 committed:
Switch from additional layer to sibling function

Switch from additional layer to sibling function; factor out
preconditions code to dedicated file; add unit tests.

matte21 commented:
> I'd introduce a sibling function there, label the existing one deprecated, and then make it call the new function you add.

@lavalamp 04c2660 does what you're suggesting, besides factoring out the preconditions code to a dedicated file and adding unit tests for the new function. I wonder whether we should also add tests to GuaranteedUpdate, the caller of the preconditions check, to have a test that fails before the fix in this PR and succeeds after it.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/storage/preconditions_test.go:
```diff
@@ -0,0 +1,192 @@
+			gotErr := tc.preconditions.SafeCheck(tc.objName, tc.obj, tc.objExists)
+			if err := pass(gotErr, tc.expectedErrorMessageWords); err != nil {
+				t.Fatal(err)
+			}
```
Can you also run the regular check method and verify that it returns the same pass / fail value (but no need to make assertions about the error it returns)? That way it will be harder for people to not update both functions.

MikeSpreitzer commented:
/retest
